kube-client/src/client/builder.rs

```diff
@@ -0,0 +1,181 @@
+use bytes::Bytes;
+use http::{Request, Response};
+use hyper::{self, client::HttpConnector};
+use hyper_timeout::TimeoutConnector;
+pub use kube_core::response::Status;
+use tower::{util::BoxService, BoxError, Layer, Service, ServiceBuilder};
+use tower_http::{
+    classify::ServerErrorsFailureClass, map_response_body::MapResponseBodyLayer, trace::TraceLayer,
+};
+
+use crate::{client::ConfigExt, Client, Config, Error, Result};
+
+/// HTTP body of a dynamic backing type.
+///
+/// The suggested implementation type is [`hyper::Body`].
+pub type DynBody = dyn http_body::Body<Data = Bytes, Error = BoxError> + Send + Unpin;
+
+/// Builder for [`Client`] instances with customized [tower](`Service`) middleware.
+pub struct ClientBuilder<Svc> {
+    service: Svc,
+    default_ns: String,
+}
+
+impl<Svc> ClientBuilder<Svc> {
+    /// Construct a [`ClientBuilder`] from scratch with a fully custom [`Service`] stack.
+    ///
+    /// This method is only intended for advanced use cases, most users will want to use [`ClientBuilder::try_from`] instead,
+    /// which provides a default stack as a starting point.
+    pub fn new(service: Svc, default_namespace: impl Into<String>) -> Self
+    where
+        Svc: Service<Request<hyper::Body>>,
+    {
+        Self {
+            service,
+            default_ns: default_namespace.into(),
+        }
+    }
+
+    /// Add a [`Layer`] to the current [`Service`] stack.
+    pub fn with_layer<L: Layer<Svc>>(self, layer: &L) -> ClientBuilder<L::Service> {
```

Comment:
Agreed in theory, but there is some other stuff I'd like to clean up before pushing people towards actually using this API too much. In particular, moving towards declaring a specific type for our request extensions rather than just piping through raw `String`s and hoping that nobody else uses that type...

Reply:
yeah, that's fair